component/kafka/component.go

```diff
@@ -0,0 +1,527 @@
+}
+
+// WithSaramaConfig specified a sarama consumer config. Use this to set consumer config or sarama level.
+// Initialize a new sarama config:
+//    c := sarama.NewConfig()
+//
+// Set consumer offset autocommit interval:
+//    c.Consumer.Offsets.AutoCommit.Enable = false
+//    c.Consumer.Offsets.AutoCommit.Interval = duration
+//
+// Set rebalance strategy
+//    c.Consumer.Group.Rebalance.Strategy = sarama.BalanceStrategySticky
+//
+// Set timeout
+//    c.Net.DialTimeout = timeout
+//
+// Set Kafka version
+//    c.Version = version
+//
+// Set offset
+//    c.Consumer.Offsets.Initial = offset
+//    c.Consumer.Offsets.Initial = sarama.OffsetOldest
+//    c.Consumer.Offsets.Initial = sarama.OffsetNewest
+//
+// Check the sarama config documentation for more config options.
+func (cb *Builder) WithSaramaConfig(cfg *sarama.Config) *Builder {
+	cb.saramaConfig = cfg
+	return cb
+}
+
+// Create constructs the Component applying
+func (cb *Builder) Create() (*Component, error) {
+	if len(cb.errors) > 0 {
+		return nil, patronErrors.Aggregate(cb.errors...)
+	}
+
+	if cb.saramaConfig == nil {
+		defaultSaramaCfg, err := defaultSaramaConfig(cb.name)
+		if err != nil {
+			return nil, err
+		}
+
+		cb.saramaConfig = defaultSaramaCfg
+	}
+
+	c := &Component{
+		name:         cb.name,
+		group:        cb.group,
+		topics:       cb.topics,
+		brokers:      cb.brokers,
+		saramaConfig: cb.saramaConfig,
+		proc:         cb.proc,
+		failStrategy: cb.failStrategy,
+		batchSize:    cb.batchSize,
+		batchTimeout: cb.batchTimeout,
+		retries:      cb.retries,
+		retryWait:    cb.retryWait,
+		commitSync:   cb.commitSync,
+	}
+
+	return c, nil
+}
+
+// Component is a kafka consumer implementation that processes messages in batch
+type Component struct {
+	name         string
+	group        string
+	topics       []string
+	brokers      []string
+	saramaConfig *sarama.Config
+	proc         BatchProcessorFunc
+	failStrategy FailStrategy
+	batchSize    uint
+	batchTimeout time.Duration
+	retries      uint
+	retryWait    time.Duration
+	commitSync   bool
+}
+
+// Run starts the consumer processing loop messages.
+func (c *Component) Run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	handler := newConsumerHandler(ctx, cancel, c.name, c.group, c.proc, c.failStrategy, c.batchSize,
+		c.batchTimeout, c.retries, c.retryWait, c.commitSync)
+	client, err := sarama.NewConsumerGroup(c.brokers, c.group, c.saramaConfig)
+	if err != nil {
+		log.Errorf("error creating consumer group client for kafka consumer component: %v", err)
+		return fmt.Errorf("error creating kafka consumer component: %w", err)
+	}
+
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			// `Consume` should be called inside an infinite loop, when a
+			// server-side rebalance happens, the consumer session will need to be
+			// recreated to get the new claims
+			if err := client.Consume(ctx, c.topics, handler); err != nil {
```

Comment:
Question: if the client error for other reasons than rebalancing, should we not recreate the client from scratch? we might have a client in a "corrupted" state. The initial kafka consumer did recreate the client from scratch.

Reply:
That's as per the sarama documentation actually. But what kind of `corrupted` state would you expect? Shouldn't that state just take the whole component down which in turn will take the whole service down and require a restart?

Reply:
The idea is to go into a retry loop by recreating everything, until the retries run out and then we fail.
The concept is that the service can stay up and does not need to be recreated every time which is especially helpful for heavy services that have a bigger warmup period. By recreating the whole client we start fresh like a new service restart. BTW The current Kafka implementation has this feature.

Reply:
Actually one of the problems of the current Kafka implementation is that it has this feature. Some problems with locks within the Sarama package (perhaps we can overcome them) But my main argument here is that not every Kafka processing error requires a full restart of the consumer. For example if someone is writing a batch to the database and the database returns a timeout. In 99% of cases either a simple retry of processing the messages will work or the consumer has hit an irrecoverable error with a message (like a poisonous message that can never be processed) which will kill the consumer.

I find it hard to understand what kind of corrupted state can a consumer be. Actually it can't because the offsets won't move; if the address of Kafka changes for example then that warrants a service restart.

If you have examples of what can be a corrupted state we can discuss further but I am very pro that the retry mechanism is on message level and not entire component level.

Reply:
You have a point here but there is one case where your current implementation is not handling it as good as the previous one and it's the case of a temp network problem. In this case the previous implementation did enter the retry state until the connectivity is up again. This implementation will exit the pod and all client calling these service will fail due to that.
Another point is that spinning up a service might have some warm up to do which means that the restart time will be even bigger.
Another point is that some services are fully ok to lose the Kafka connection for some time and they can serve even with that situation for some time.
The cases mentioned are not fictitious. Sonar actually relies on the driver position which we might be able to serve some stale data until the Kafka connection is repaired. Sonar also consumes the DMV topic where we get updates rarely so we can still work with "stale" data as the impact of this situation is rides with not an up to date service. 
Sonar has a warm up period of around 10-30 sec. When a network problem happens, the current Kafka component will retry to create all the connectivity from scratch without restarting the service. During this period Sonar serves requests as usual so the service availability is there.
In this implementation, the service would restart, it would stay in a restart state until the Kafka connectivity is up again and then take another 10-30 sec to warmup. This this time would be a full downtime for Sonar in comparison to the previous implementation where the service would be up the whole time.
To summarize what we lose without these retries:
- Service termination and warm up on restart will leave the service not available for a longer period (service availability)
- Stale data can be fine in some scenarios

As for problems regarding the errors returned from the process function. Obviously not being able to process a message should not terminate the service. Usually if the message cannot be processed we should no mark the messages at all and let the service handle them again and again until the underlying problem is fixed. That is the reason we should give full control of the marking of the messages to the developers in the proc function so they can do it.
The idea was that the Kafka component should be as "stupid" as possible and give as much freedom and full control as possible so that the end user takes the responsibility of his actions and decides to mark or net the messages.

Hope this help understand the situation better.

Reply:
> the case of a temp network problem

Shouldn't we rely on Sarama's configuration for this? There is a bit of logic there to handle these, if I am not mistaken, and I believe Sarama has better context about network problems because it manages the connection pool and the brokers individually.



> The idea was that the Kafka component should be as "stupid" as possible and give as much freedom and full control as possible so that the end user takes the responsibility of his actions and decides to mark or net the messages.

This is the best way to handle it :+1: we strike a good balance here with Patron's interfaces.

I am sorry if it sounds stupid but the best "flavor" that I would like for this new component would be one without any exit strategy at all: Go has channels and (the dreaded) `context.Context` for this type of coordination, that we already use.

On the other hand I am respectful for all the work @oss92 invested here to add the new exit strategies and I have no doubt they have fitting use cases in existing services as @mantzas mentioned, so perhaps it is just my view too simplistic.

Reply:
As agreed, retries now restart the whole component. Retries on message level need to be managed by clients.